ethereum/contracts/relayer/coreRelayer/CoreRelayerSerde.sol

```diff
@@ -0,0 +1,272 @@
+// SPDX-License-Identifier: Apache 2
+
+pragma solidity ^0.8.19;
+
+import {
+  InvalidPayloadId,
+  InvalidPayloadLength,
+  InvalidVaaKeyType,
+  VaaKey,
+  VaaKeyType,
+  Send,
+  ExecutionParameters,
+  DeliveryInstruction,
+  RedeliveryInstruction,
+  DeliveryOverride
+} from "../../interfaces/relayer/IWormholeRelayer.sol";
+import {BytesParsing} from "./BytesParsing.sol";
+
+library CoreRelayerSerde {
+  using BytesParsing for bytes;
+
+  // ---------------------- "public" (i.e implicitly internal) encode/decode -----------------------
+
+  uint8 private constant PAYLOAD_ID_VAAKEY = 1;
+  uint8 private constant PAYLOAD_ID_EXECUTION_PARAMETERS = 1;
+  uint8 private constant PAYLOAD_ID_DELIVERY_INSTRUCTION = 1;
+  uint8 private constant PAYLOAD_ID_DELIVERY_OVERRIDE = 1;
+  //TODO AMO: Why are we at version 2 here?
+  uint8 private constant PAYLOAD_ID_REDELIVERY_INSTRUCTION = 2;
+
+  //TODO AMO: All the recursive abi.encodePacked calls in here are _insanely_ gas inefficient unless
```

Comment:
Ah the only reason for this was stack-too-deep related reasons, which are alleviated now that we have via-ir.

Reply:
I took a stab at removing some in https://github.com/wormhole-foundation/wormhole/pull/2899, but all the ones related to encoding arrays seem tough to remove

Reply:
I think we can't encode arrays efficiently with the current expressive power of solidity. We always end with a quadratic cost in terms of memory. At some point I wanted to write some assembly to optimize this part and concatenate elements in a linear algorithm but I think it can be done after the first release.